services/ext/service.go

```diff
@@ -645,24 +644,51 @@ func (s *Service) FetchCommunityInfo(communityID string) (*thirdparty.CommunityI
 	return communityToInfo(community), nil
 }
 
+func (s *Service) FetchCommunityInfoForCollectibles(communityID string, ids []thirdparty.CollectibleUniqueID) (*thirdparty.CommunityInfo, error) {
+	community, err := s.messenger.FindCommunityInfoFromDB(communityID)
+	if err != nil && err != communities.ErrOrgNotFound {
+		return nil, err
+	}
+
+	if community == nil {
+		return s.FetchCommunityInfo(communityID)
+	}
+
+	hasAllMetadata := true
+	for _, id := range ids {
+		tokenMetadata, err := s.fetchCommunityCollectibleMetadata(community, id.ContractID)
+		if err != nil {
+			return nil, err
+		}
+		if tokenMetadata == nil {
+			hasAllMetadata = false
+			break
+		}
+	}
+
+	if !hasAllMetadata {
+		return s.FetchCommunityInfo(communityID)
+	}
+
+	return communityToInfo(community), nil
+}
+
 func (s *Service) fetchCommunity(communityID string, fetchLatest bool) (*communities.Community, error) {
 	if s.messenger == nil {
 		return nil, fmt.Errorf("messenger not ready")
 	}
 
-	// Try to fetch metadata from Messenger communities
-
-	// TODO: we need the shard information in the collectible to be able to retrieve info for
-	// communities that have specific shards
+	// Get the latest successfully fetched version of the Community
+	community, err := s.messenger.FindCommunityInfoFromDB(communityID)
+	if err != nil && err != communities.ErrOrgNotFound {
+		return nil, err
+	}
 
-	if fetchLatest {
-		// Try to fetch the latest version of the Community
-		var shard *shard.Shard = nil // TODO: build this with info from token
-		// NOTE: The community returned by this function will be nil if
-		// the version we have in the DB is the latest available.
-		_, err := s.messenger.FetchCommunity(&protocol.FetchCommunityRequest{
+	// Don't fetch from store nodes if we are a control node,
+	// we already have the latest version in the db.
+	if !(community != nil && community.IsControlNode()) && fetchLatest {
```

Comment:
I think we need to always fetch if `community == nil`. Here, if the community is `nil` but  not `fetchLatest`, we'll return `nil`. I'm not sure if that's ok

Reply:
It is the same behavior as before.

Reply:
yeah perhaps the function's / boolean's name were already sub-optimal.
With the new logic, the most reasonable way to arrange this might be do away with the boolean parameter and simply call this (renamed to `fetchCommunityIfNeeded`?) or `getCommunityFromDB` where needed. But I'll let you guys decide since this is messenger territory.